File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    var set = new Set()
    i = 0, j = 0, max = 0
    while(j < s.length){
        if(!set.has(s[j])){
            set.add(s[j++])
            max = Math.max(max, set.size)
        } else {
            set.delete(s[i++])
        }
    }
    return max
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        int l = 0;
        int r = 0;
        int max = 0;
        
        char[] chars = s.toCharArray();
        while(r < chars.length){
            if(!set.contains(chars[r])){
                set.add(chars[r]);
                r++;
            }
            else {
                max = Math.max(max, set.size());
                set.remove(chars[l]);
                l++;
            }
        }
        return Math.max(max, set.size());
    }
}